compiler/sem.nim

```diff
@@ -158,11 +158,12 @@ proc commonType*(x, y: PType): PType =
       a = a.lastSon.skipTypes({tyGenericInst})
       b = b.lastSon.skipTypes({tyGenericInst})
     if a.kind == tyObject and b.kind == tyObject:
-      result = commonSuperclass(a, b, k)
+      result = commonSuperclass(a, b)
       # this will trigger an error later:
       if result.isNil or result == a: return x
       if result == b: return y
-      if k != tyNone:
+      # bug #7906
+      if k != tyNone and x.kind != tyGenericInst:
```

Comment:
Please add a comment here explaining your reasoning. It seems fishy, why is ``tyGenericInst`` harmful here?

Reply:
I will add a comment.

the reason is
```Nim
type
   GenericObject[T] = object
   GenericRefObject[T] = ref object
   
var a = GenericRefObject[int]          # tyGenericInst of ref object
var b = new(ref GenericObject[int]) # tyRef of tyGenericInst
```

from user perpective, a and b should be not too different, but inside the compiler, they are two different types.

so, if it already a tyGenericInst of ref object, it would be wrong to add more tyRef in front of it.
additional tyRef in the next compiler phase will generate ill formed AST.